packages/router/src/router.tsx

```diff
@@ -200,77 +190,64 @@ const LocationAwareRouter: React.FC<RouterProps> = ({
+    id: number
+    wrappers: Wrappers
+    isPrivate: boolean
+    props: {
+      private?: boolean
+      [key: string]: unknown
+    }
+  }>
 }
 
 /**
  * This is effectively a Set (without auth-related code)
  *
- * This means that the <Set> and <Private> components become "virtual"
+ * This means that the <Set> and <PrivateSet> components become "virtual"
  * i.e. they are never actually Rendered, but their props are extracted by the
  * analyze routes function.
  *
  * This is so that we can have all the information up front in the routes-manifest
  * for SSR, but also so that we only do one loop of all the Routes.
  */
-const WrappedPage = memo(
-  ({ wrappers, routeLoaderElement, setProps }: WrappedPageProps) => {
-    // @NOTE: don't mutate the wrappers array, it causes full page re-renders
-    // Instead just create a new array with the AuthenticatedRoute wrapper
-
-    // we need to pass the setProps from each set to each wrapper
-    let wrappersWithAuthMaybe = wrappers
-    const reveresedSetProps = [...setProps].reverse()
-
-    reveresedSetProps
-      // @MARK note the reverse() here, because we spread wrappersWithAuthMaybe
-      .forEach((propsFromSet) => {
-        if (propsFromSet.private) {
-          if (!propsFromSet.unauthenticated) {
-            throw new Error(
-              'You must specify an `unauthenticated` route when marking a Route as private'
-            )
-          }
-
-          // @MARK: this component intentionally removes all props except children
-          // because the .reduce below will apply props inside out
-          const AuthComponent: React.FC<{ children: ReactNode }> = ({
-            children,
-          }) => {
-            return (
-              <AuthenticatedRoute {...propsFromSet}>
-                {children}
-              </AuthenticatedRoute>
-            )
-          }
+const WrappedPage = memo(({ routeLoaderElement, sets }: WrappedPageProps) => {
+  // @NOTE: don't mutate the wrappers array, it causes full page re-renders
+  // Instead just create a new array with the AuthenticatedRoute wrapper
 
-          wrappersWithAuthMaybe = [AuthComponent, ...wrappersWithAuthMaybe]
-        }
-      })
+  if (!sets || sets.length === 0) {
+    return routeLoaderElement
+  }
 
-    if (wrappersWithAuthMaybe.length > 0) {
-      // If wrappers exist e.g. [a,b,c] -> <a><b><c><routeLoader></c></b></a> and returns a single ReactNode
-      // Wrap AuthenticatedRoute this way, because if we mutate the wrappers array itself
-      // it causes full rerenders of the page
-      return wrappersWithAuthMaybe.reduceRight((acc, wrapper) => {
-        // Merge props from set, the lowest set props will override the higher ones
-        const mergedSetProps = setProps.reduce((acc, props) => {
-          return { ...acc, ...props }
-        }, {})
+  return sets.reduceRight<ReactNode | undefined>((acc, set) => {
+    // Bundle up all the wrappers into a single element with each wrapper as a
```

Comment:
Mind leaving this comment here? 
```
If wrappers exist e.g. [a,b,c] -> <a><b><c><routeLoader></c></b></a> and returns a single ReactNode
```

Everytime I look at this code, I forget. My brain doesn't handle reduce.

Reply:
One of the many things I appreciate about our code is how easy it is to read and understand. Our generous use of comments is a big part of that. Of course I'll add it back in!